### PR TITLE
Initial implementation of anonymous_pipe API

### PR DESCRIPTION
**Pull request by NobodyXu**
ACP completed in rust-lang/libs-team#375
Tracking issue: #127154

try-job: x86_64-msvc
try-job: i686-mingw

### CODE REVIEW

File: library/std/src/net/pipe/mod.rs
Comment (NobodyXu) [inline library/std/src/net/pipe/mod.rs:139]: Not sure if it is the right places as it push hook warns me to not use `cfg(unix)` here.

I checked and it seems that `AsFd` implementation is put inside `owned_fd` mod.

However I don't think I can put `TryFrom<OwnedFd> for PipeReader` in there

Comment (NobodyXu) [reply library/std/src/net/pipe/mod.rs]: Figured it out, moving it to `src/sys` fixed it

File: library/std/src/sys/pal/unix/fd.rs
Comment (NobodyXu) [reply library/std/src/sys/pal/unix/fd.rs]: Thanks updated

File: library/std/src/sys/pal/unix/fd.rs
Comment (jhpratt) [inline library/std/src/sys/pal/unix/fd.rs:550]: My thinking was to avoid this altogether. Is it ever actually reachable?

Comment (NobodyXu) [reply library/std/src/sys/pal/unix/fd.rs:550]: According to the [`open`](https://man7.org/linux/man-pages/man2/open.2.html) it should not happen, however I think it should not panic upon any unexpected value.

I think we can fallback to read-only in that case?

File: library/std/src/net/mod.rs
Comment (a1phyr) [inline library/std/src/net/mod.rs:37]: In the ACP this was in a new `pipe` module under `std::io`. `std::net` seems fine to me but this should maybe have its own `pipe` module anyway ?

File: library/std/src/sys/anonymous_pipe/mod.rs
Comment (a1phyr) [inline library/std/src/sys/anonymous_pipe/mod.rs:13]:
```suggestion
            Err(io::Error::UNSUPPORTED_PLATFORM)
```

File: library/std/src/sys/anonymous_pipe/mod.rs
Comment (a1phyr) [inline library/std/src/sys/anonymous_pipe/mod.rs:68]: Using a macro for two impls is more harmful than helpful and not consistent with what is done elsewhere in `std` anyway.

(Same for `forward_io_write_traits!` below and `impl_traits!` in other files)

File: library/std/src/sys/anonymous_pipe/unix.rs
Comment (a1phyr) [inline library/std/src/sys/anonymous_pipe/unix.rs:74]: `File` implements `From<OwnedFd>` directly without any check. Should we do the same here for consistency ? 

Same question for `PipeWriter` and for `From<OwnedHandle>` on Windows.

File: library/std/src/sys/pal/unix/fd.rs
Comment (a1phyr) [reply library/std/src/sys/pal/unix/fd.rs:550]: Returning an error seems the right thing to me.

File: library/std/src/net/mod.rs
Comment (NobodyXu) [reply library/std/src/net/mod.rs:37]: Yeah I can move it to a separate module.

File: library/std/src/sys/anonymous_pipe/unix.rs
Comment (NobodyXu) [reply library/std/src/sys/anonymous_pipe/unix.rs:74]: Hmmm, yeah perhaps it makes sense to remove it.

It does not pose any memory safety issue whatever, I can remove it

Comment (NobodyXu) [reply library/std/src/sys/anonymous_pipe/unix.rs]: I've removed it.

File: library/std/src/sys/anonymous_pipe/mod.rs
Comment (NobodyXu) [reply library/std/src/sys/anonymous_pipe/mod.rs]: Removed.

File: library/std/src/net/mod.rs
Comment (NobodyXu) [reply library/std/src/net/mod.rs]: Moved it to new module `std::pipe`

File: library/std/src/sys/pal/unix/fd.rs
Comment (NobodyXu) [reply library/std/src/sys/pal/unix/fd.rs]: Replaced `TryFrom<Owned*>` with `From` to maintain consistency with `File`

https://github.com/rust-lang/rust/pull/127153#discussion_r1672170703

**Comment by jhpratt** [comment]
LGTM. Thanks @a1phyr for the additional input!

@bors r=jhpratt,a1phyr rollup

**Comment by jhpratt** [comment]
@bors r-

Failing in CI: https://github.com/rust-lang/rust/pull/127597#issuecomment-2222178347

**Comment by NobodyXu** [comment]
cc @jhpratt I pushed a fix for the compilation error on unsupported platforms.

**Comment by matthiaskrgr** [comment]
Ceci n'est pas une pipe.

@bors r- 

https://github.com/rust-lang-ci/rust/actions/runs/9907462928/job/27371229875#step:25:3177

**Comment by NobodyXu** [comment]
Thanks, I pushed a fix again.

This time I successfully run compilation for wasm32-unknown-unknown (previously failed due to not setting wasi root)

cc @jhpratt

**Comment by jhpratt** [comment]
Let's try this one more time

@bors r+

**Comment by workingjubilee** [comment]
I believe this has failed in #127701 see https://github.com/rust-lang/rust/pull/127701#issuecomment-2227175515

@bors r- rollup=iffy

File: library/std/src/sys/anonymous_pipe/mod.rs
Comment (workingjubilee) [inline library/std/src/sys/anonymous_pipe/mod.rs:1]:
```suggestion
//! Module for anonymous pipe
```

File: library/std/src/sys/anonymous_pipe/mod.rs
Comment (workingjubilee) [inline library/std/src/sys/anonymous_pipe/mod.rs:13]:
```suggestion
/// Create anonymous pipe that is close-on-exec and blocking.
```

File: library/std/src/sys/anonymous_pipe/mod.rs
Comment (workingjubilee) [inline library/std/src/sys/anonymous_pipe/mod.rs:28]:
```suggestion
/// Read end of the anonymous pipe.
```

File: library/std/src/sys/anonymous_pipe/mod.rs
Comment (workingjubilee) [inline library/std/src/sys/anonymous_pipe/mod.rs:33]:
```suggestion
/// Write end of the anonymous pipe.
```

**Review by workingjubilee — COMMENTED** [review]
nit: typos

**Comment by NobodyXu** [comment]
> I believe this has failed in #127701 see https://github.com/rust-lang/rust/pull/127701#issuecomment-2227175515


seems that miri doesn't support pipe2, so tests should not be run on miri.

```
  error: unsupported operation: can't call foreign function `pipe2` on OS `linux`
  Error:     --> library/std/src/sys/pal/unix/pipe.rs:32:21
       |
  32   |                 cvt(libc::pipe2(fds.as_mut_ptr(), libc::O_CLOEXEC))?;
       |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can't call foreign function `pipe2` on OS `linux`
       |
       = help: if this is a basic API commonly used on this target, please report an issue with Miri
       = help: however, note that Miri does not aim to support every FFI function out there; for instance, we will not support APIs for things such as GUIs, scripting languages, or databases
       = note: BACKTRACE on thread `sys::anonymous_`:
       = note: inside `sys::pal::unix::pipe::anon_pipe` at library/std/src/sys/pal/unix/pipe.rs:32:21: 32:67
  note: inside `sys::anonymous_pipe::unix::pipe`
      --> library/std/src/sys/anonymous_pipe/unix.rs:15:5
       |
  15   |     anon_pipe().map(|(rx, tx)| (PipeReader(rx), PipeWriter(tx)))
       |     ^^^^^^^^^^^
  note: inside `sys::anonymous_pipe::pipe`
      --> library/std/src/sys/anonymous_pipe/mod.rs:19:13
       |
  19   |             unix::pipe()
       |             ^^^^^^^^^^^^
  note: inside `sys::anonymous_pipe::tests::pipe_creation_and_rw`
      --> library/std/src/sys/anonymous_pipe/tests.rs:6:28
       |
  6    |     let (mut rx, mut tx) = pipe().unwrap();
       |                            ^^^^^^
  note: inside closure
      --> library/std/src/sys/anonymous_pipe/tests.rs:5:26
       |
  4    | #[test]
       | ------- in this procedural macro expansion
  5    | fn pipe_creation_and_rw() {
       |                          ^
       |
      ::: /checkout/library/core/src/macros/mod.rs:1641:5
       |
  1641 |     pub macro test($item:item) {
       |     -------------- in this expansion of `#[test]`
  note: inside `<{closure@library/std/src/sys/anonymous_pipe/tests.rs:5:1: 13:2} as core::ops::FnOnce<()>>::call_once - shim`
      --> /checkout/library/core/src/ops/function.rs:250:5
       |
  250  |     extern "rust-call" fn call_once(self, args: Args) -> Self::Output;
       |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  note: inside `<fn() -> core::result::Result<(), realstd::string::String> as core::ops::FnOnce<()>>::call_once - shim(fn() -> core::result::Result<(), realstd::string::String>)`
      --> /checkout/library/core/src/ops/function.rs:250:5
       |
  250  |     extern "rust-call" fn call_once(self, args: Args) -> Self::Output;
       |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  
  note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace
  
  error: aborting due to 1 previous error
  
  error: test failed, to rerun pass `-p std --lib`
  
```

File: library/std/src/sys/anonymous_pipe/mod.rs
Comment (NobodyXu) [inline library/std/src/sys/anonymous_pipe/mod.rs:140]:
```suggestion
#[cfg(all(test, not(miri)))]
```

**Comment by NobodyXu** [comment]
cc @jhpratt can we try again please?

**Comment by jhpratt** [comment]
Would you mind cleaning up the commit history some? There's a number of commits that can be squashed in an interactive rebase. A quick look shows that perhaps the five most recent would be those?

**Comment by NobodyXu** [comment]
Sure, will do it

**Comment by jhpratt** [comment]
@bors delegate+

`r=me` when that's done. Hopefully this should be the final change necessary :crossed_fingers:

**Comment by NobodyXu** [comment]
Fixed the error, let's try again

@bors r=@jhpratt

**Comment by tgross35** [comment]
@NobodyXu if you're around this could use a squash, the commits make some changes then revert them (nbd if you don't get to it before bors does)

**Comment by tgross35** [comment]
@bors r-

Actually please do squash, it looks like the cleanup requested at https://github.com/rust-lang/rust/pull/127153#issuecomment-2227205667 must have been lost. At least remove the redundant ones and fixups, or even just squish into one.

**Comment by NobodyXu** [comment]
cc @tgross35 Squashed all commits into one

**Comment by tgross35** [comment]
Hm the x86_64 and i686 Windows jobs have been running for nearly 3 hours, usual is about 1:30. I'll give it about 10 more minutes before canceling, but that might need a closer look (cc @ChrisDenton).

**Comment by tgross35** [comment]
@bors r-
@bors rollup=never

I set this PR's try jobs to two of the failing tests, once you have a fix run `@bors try` to see if they complete

**Comment by ChrisDenton** [comment]
The internal pipe functions were not really designed for in-process use and definitely not for single-threaded. They were designed so that `ours` is async and kept and `theirs` is sync and given to another process. I think, at a minimum, `PipeWriter` should just wrap a normal `Handle`, i.e. `PipeWriter(Handle);`. And the test should spawn a thread for writing.

**Comment by NobodyXu** [comment]
Thanks

> I think, at a minimum, PipeWriter should just wrap a normal Handle, i.e. PipeWriter(Handle);.

`Handle` doesn't have read/write implementation though, so we'd have to keep the current one.

> And the test should spawn a thread for writing.

Thanks, in terms of API, is it possible to make them usable within one thread?

Because we are providing a cross-platform API here, if the `pipe()` only works on unix then it's not really cross-platform.

**Comment by tgross35** [comment]
os_pipe seems to use the handle as a file to get the read/write impls https://github.com/oconnor663/os_pipe.rs/blob/6268f8639904dcb6adb461fbdbcda959e4eaffe2/src/windows.rs#L84-L88, is this robust?

Cc @oconnor663 (maintainer) if you have any input here.

**Comment by ChrisDenton** [comment]
I'm not sure that solving the single-threaded case with synchronous writes is entirely possible. There's always a risk that `WriteFile` will block indefinitely (e.g. writing when the buffer is fall) and then that will deadlock.

**Comment by ChrisDenton** [comment]
Though to be honest, I don't understand the use case of a thread sending a message to itself.

**Comment by NobodyXu** [comment]
> Though to be honest, I don't understand the use case of a thread sending a message to itself.

Yeah that doesn't make sense, so I will update the test to spawn a thread for writer.

**Comment by ChrisDenton** [comment]
I think it might be worth de-coupling the handles we make for `Command` and the ones we make for `pipe` instead of reusing them. We don't directly expose any async read/write in the standard library so the `pipe` case can be greatly simplified to just using [`CreatePipe`](https://learn.microsoft.com/en-us/windows/win32/api/namedpipeapi/nf-namedpipeapi-createpipe) and normal `File` read/write operations. This will avoid a lot of issues with wrangling named pipes.

**Comment by NobodyXu** [comment]
> We don't directly expose any async read/write in the standard library so the `pipe` case can be greatly simplified to just using [`CreatePipe`](https://learn.microsoft.com/en-us/windows/win32/api/namedpipeapi/nf-namedpipeapi-createpipe)

I was thinking about it, but then the os-pipe says they want to switch to named pipes:

https://github.com/oconnor663/os_pipe.rs/blob/6268f8639904dcb6adb461fbdbcda959e4eaffe2/src/windows.rs#L15

While stdlib does not use pipe for async I/O, maybe libraries using it would want to use it?

I.e. if tokio/async-std decides to use stdlib pipe API for async operations?

**Comment by ChrisDenton** [comment]
I think that should be an extension API rather than in the initial implementation. Currently we have the arbitrary situation where one handle is async and the other is sync, with this fact being hidden and not obvious.

**Comment by NobodyXu** [comment]
> I think that should be an extension API rather than in the initial implementation. Currently we have the arbitrary situation where one handle is async and the other is sync, with this fact being hidden and not obvious.

Yeah that makes sense, I'd overhaul the windows pipe implementation.

**Comment by NobodyXu** [comment]
Updated PR, windows now use an unnamed pipe creation.

**Comment by oconnor663** [comment]
> os_pipe seems to use the handle as a `File` to get the read/write impls https://github.com/oconnor663/os_pipe.rs/blob/6268f8639904dcb6adb461fbdbcda959e4eaffe2/src/windows.rs#L84-L88, is this robust?
> 
> Cc @oconnor663 (maintainer) if you have any input here.

Yeah I've always felt a little guilty about this. Since the standard library already has a pipe abstraction on the inside, it makes perfect sense to use that instead of `File`. That said, since the standard library also _owns_ `File`, it would also be less sketchy to assume its implementation details :)

> I was thinking about it, but then the os-pipe says they want to switch to named pipes:

For what it's worth, that comment has been there since 2016 :sweat_smile: `os_pipe` doesn't provide any async IO integration, and `tokio` has its own pipe abstractions internally (I think they have to do it themselves to register wakeups with their runtime?), so it's unlikely that `os_pipe` will ever switch.

**Comment by NobodyXu** [comment]
Thanks!

> tokio has its own pipe abstractions internally (I think they have to do it themselves to register wakeups with their runtime?)

Turns out tokio only provide the [pipe](https://docs.rs/tokio/latest/tokio/net/unix/pipe/index.html) on UNIX, and IIRC their process implementation does some special stuff with std::process stdio on windowe to get the async read.

**Comment by oconnor663** [comment]
By the way, there's a typo in the commit message, the word "anonymous".

**Comment by NobodyXu** [comment]
> By the way, there's a typo in the commit message, the word "anonymous".

Thanks, fixed

File: library/std/src/sys/anonymous_pipe/windows.rs
Comment (NobodyXu) [reply library/std/src/sys/anonymous_pipe/windows.rs]: Updated

File: library/std/tests/pipe_subprocess.rs
Comment (NobodyXu) [reply library/std/tests/pipe_subprocess.rs:3]: Updated

**Comment by NobodyXu** [comment]
cc @tgross35 @ChrisDenton CI is now green

**Comment by NirjharRoyiitk** [comment]
two tests fail on x86_64-fortanix-unknown-sgx target i.e.
```
02:57:13 ---- sys::anonymous_pipe::tests::pipe_creation_and_rw stdout ----
02:57:13 thread 'sys::anonymous_pipe::tests::pipe_creation_and_rw' panicked at library/std/src/sys/anonymous_pipe/tests.rs:6:35:
02:57:13 called `Result::unwrap()` on an `Err` value: Error { kind: Unsupported, message: "operation not supported on this platform" }
02:57:13 
02:57:13 ---- sys::anonymous_pipe::tests::pipe_try_clone_and_rw stdout ----
02:57:13 thread 'sys::anonymous_pipe::tests::pipe_try_clone_and_rw' panicked at library/std/src/sys/anonymous_pipe/tests.rs:17:35:
02:57:13 called `Result::unwrap()` on an `Err` value: Error { kind: Unsupported, message: "operation not supported on this platform" }
```

**Comment by NobodyXu** [comment]
Thanks @NirjharRoyiitk , is [x86_64-fortanix-unknown-sgx](https://doc.rust-lang.org/rustc/platform-support/x86_64-fortanix-unknown-sgx.html#x86_64-fortanix-unknown-sgx) a unix platform?

The tests are only run on unix and windows and is also disabled when running miri

**Comment by raoulstrackx** [comment]
> Thanks @NirjharRoyiitk , is [x86_64-fortanix-unknown-sgx](https://doc.rust-lang.org/rustc/platform-support/x86_64-fortanix-unknown-sgx.html#x86_64-fortanix-unknown-sgx) a unix platform?
No SGX doesn't run on top of an OS. It uses CPU features to run enclaves in complete isolation of the rest of the system. Many things like pipes don't make much sense for this platform.

> The tests are only run on unix and windows and is also disabled when running miri
Sounds great! Thanks!

**Comment by NobodyXu** [comment]
> No SGX doesn't run on top of an OS. It uses CPU features to run enclaves in complete isolation of the rest of the system. Many things like pipes don't make much sense for this platform.

Thanks, in that case the tests should be disabled.

@NirjharRoyiitk I just checked the output again and it seems that you are running an older version of this PR, the latest one should have the tests disabled for your target, can you try it again please?

**Comment by NobodyXu** [comment]
Damn I forgot to check in the new files I created 😂

@bors r=@jhpratt

**Comment by tgross35** [comment]
@NobodyXu given Chris's concerns about the feasibility of supporting the Windows interface, and the fact that the implementation has changed significantly since Jacob last took a look at it, it would be best to wait for a re-review.

There is no rush here, it will likely be at the top of the queue whenever it is approved.

@bors r-
r? @ChrisDenton

File: library/std/src/sys/anonymous_pipe/mod.rs
Comment (ChrisDenton) [inline library/std/src/sys/anonymous_pipe/mod.rs:1]: The public docs and types should be in a top level module `src/pipe.rs` rather than in sys.

**Review by ChrisDenton — COMMENTED** [review]
I'm ok with the code changes but I do have one issue.

File: library/std/src/sys/anonymous_pipe/mod.rs
Comment (NobodyXu) [reply library/std/src/sys/anonymous_pipe/mod.rs:1]: Yep I could do that, but I have to leave the FromRawFd impl in the sys, because only sys allows use of `cfg(unix)`

Comment (NobodyXu) [reply library/std/src/sys/anonymous_pipe/mod.rs:1]: I would then have to leave the unix/windowe/unsupported and the test module in sys, while moving pipe into src

Is there any better filesystem layout?

Comment (ChrisDenton) [reply library/std/src/sys/anonymous_pipe/mod.rs:1]: `sys` should expose a single interface across all platforms so in this case it just needs `AnonPipe` and `pipe`. This makes it obvious what platforms have to provide and means that people working in the non-sys parts of std don't have to wade into platform specific details. Keeping things in `sys/anonymous_pipe.rs` is fine.

**Comment by NobodyXu** [comment]
@bors try

Just to make sure that everything works fine

**Review by ChrisDenton — APPROVED** [review]
Thanks

**Pull request merged**
